### PR TITLE
refactor: Convert form-feedback.client.factory.js, CsvGenerator.js to TypeScript

### LINKED ISSUE #1509: Convert form-feedback.client.factory.js, CsvGenerator.js to TypeScript

### PR DESCRIPTION
**Pull request by tshuli**
## Problem
Closes #1509 

## Tests

- [ ] Create email mode form. Check that feedback tab shows no feedback received.
- [ ] Check that form feedback can be submitted successfully. Do this several times.
- [ ] Check that submitted feedback and count appears correctly on feedback tab.
- [ ] Check that submitted feedback can be downloaded.
- [ ] Repeat above with storage mode form.

### CODE REVIEW

File: src/types/form_feedback.ts [hunk 1]
```diff
@@ -2,16 +2,43 @@ import { Document, Model, QueryCursor } from 'mongoose'
...
+export type FormFeedbackPostDto = Omit<IFormFeedback, 'formId'> & {
+  isPreview?: boolean
+}
+
+export type FormFeedbackResponseDto = IFormFeedback & {
+  created?: Date
+  lastModified?: Date
+}
+
+export type GetFormFeedbackDto = FeedbackResponse
```
Comment (mantariksh) [inline src/types/form_feedback.ts:35]: is there a need to alias this, or can we just use `GetFormFeedbackDto`?

File: src/public/modules/forms/base/components/feedback-form.client.component.js [hunk 2]
```diff
@@ -31,19 +33,14 @@ function feedbackController(FormFeedback, Toastr) {
...
-          formId: vm.formId,
-        },
-        feedback,
-      ).then(
+      FormFeedback.postFeedback(vm.formId, feedback).then(
         function (_response) {
           vm.isSubmitted = true
           vm.isLoading = false
           Toastr.success('Thank you for your submission!')
+          $scope.$apply()
```
Comment (mantariksh) [inline src/public/modules/forms/base/components/feedback-form.client.component.js:41]: if we use `$scope.$apply()` here, do we need to do so in the error handling as well? I'm not sure - if that's the case then maybe we should just stick to `$q.when`?

File: src/public/modules/forms/services/FormFeedbackService.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  return axios
+    .get<number>(`${ADMIN_FORM_ENDPOINT}/${formId}/feedback/count`)
+    .then(({ data }) => data)
+}
+
+/**
+ * Class to encapsulate the FeedbackCsv and its attributes
+ * Exported for testing
+ */
+export class FeedbackCsvGenerator extends CsvGenerator {
```
Comment (karrui) [inline src/public/modules/forms/services/FormFeedbackService.ts:61]: maybe we should move this to its own file? feels weird to have this in a service file. maybe its own `FeedbackCsvGenerator.ts` file?

File: src/public/modules/forms/services/__tests__/FormFeedbackService.spec.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+      rating: 3,
+      comment: 'some comment',
+    }
+
+    it('should call csvGenerator.addLineFromFeedback and csvGenerator.triggerFileDownloadwhen the GET request succeeds', async () => {
+      // Arrange
+      MockAxios.get
+        .mockResolvedValueOnce({ data: 1 })
+        .mockResolvedValueOnce({ data: [mockFeedback] })
+      FormFeedbackService.FeedbackCsvGenerator.prototype.addLineFromFeedback = jest.fn()
```
Comment (karrui) [inline src/public/modules/forms/services/__tests__/FormFeedbackService.spec.ts:138]: will be much easier to mock if feedbackcsvgenerator is in another file too

File: src/public/modules/forms/helpers/CsvGenerator.ts [hunk 5]
```diff
@@ -38,25 +47,25 @@ module.exports = class CsvGenerator {
    * Insert headers into the CSV file after the meta-data
    * @param {Array} headerLabels array of labels for header row
    */
-  setHeader(headerLabels) {
+  setHeader(headerLabels: (number | string)[]): void {
```
Comment (mantariksh) [inline src/public/modules/forms/helpers/CsvGenerator.ts:50]: I think the headers can only be strings, right?

File: src/public/modules/forms/services/__tests__/FormFeedbackService.spec.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
+import { ObjectId } from 'bson-ext'
```
Comment (karrui) [inline src/public/modules/forms/services/__tests__/FormFeedbackService.spec.ts:1]: there should also be tests for `FeedbackCsvGenerator` itself, since it is an utility class and behaviours are expected. No mocks should be needed. May want to see how `CsvMergedHeadersGenerator` (man we really should rename that file to something better) is tested

**Review by karrui — CHANGES_REQUESTED** [review]

**Comment by karrui** [comment]
Oh, also add tests for the base `CsvGenerator` class?

File: src/public/modules/forms/helpers/CsvGenerator.ts [hunk 7]
```diff
@@ -38,25 +47,25 @@ module.exports = class CsvGenerator {
...
     this.records[this.startIdx - 1] = stringify(headerLabels)
   }
 
   /**
    * Insert meta-data array into the start of the CSV file
    * @param {Array} metaDataRows array of arrays, metaDataRows[i][j] holds the data for row i, col j of the metaData table
    */
-  addMetaData(metaDataRows) {
+  addMetaData(metaDataRows: (number | string)[]): void {
```
Comment (mantariksh) [inline src/public/modules/forms/helpers/CsvGenerator.ts:58]: I'm not sure this typing is correct. the JSdocs says "array of arrays", and I can see e.g. from `CsvMergedHeadersGenerator.js` that the argument being passed is of type `(number | string)[][]`

File: src/app/modules/form/public-form/public-form.controller.ts [hunk 8]
```diff
@@ -120,31 +120,27 @@ export const submitFormFeedback: RequestHandler<
...
-    logger.error({
-      message: 'Error creating form feedback',
-      meta: {
-        action: 'handleSubmitFeedback',
-        ...createReqMeta(req),
-        formId,
-      },
-      error,
+    .map((submitFeedbackResult) =>
+      res.status(StatusCodes.OK).json(submitFeedbackResult),
```
Comment (mantariksh) [inline src/app/modules/form/public-form/public-form.controller.ts:129]: let's not call this variable `Result`, since one might mistake it for a `neverthrow` `Result`, which it's not

File: src/public/modules/forms/services/FormFeedbackService.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
+import axios from 'axios'
```
Comment (mantariksh) [inline src/public/modules/forms/services/FormFeedbackService.ts:1]: can we move this file into `src/public/services`? all the new services are being added there

File: src/public/modules/forms/services/FormFeedbackService.ts (hunk 3 above)
Comment (mantariksh) [inline src/public/modules/forms/services/FormFeedbackService.ts:61]: should we move this class into a separate file? all the other functions have to do with API request/response, so this seems unrelated

File: src/public/modules/forms/services/FormFeedbackService.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+ * @param formId the id of the form to download feedback for
+ * @returns a stream of feedback
+ */
+export const downloadFeedback = async (
+  formId: string,
+  formTitle: string,
+): Promise<void> => {
+  const expectedNumResponses = await countFeedback(formId)
+
+  return axios
```
Comment (mantariksh) [inline src/public/modules/forms/services/FormFeedbackService.ts:94]: do you think `async`/`await` syntax might make this clearer?

File: src/public/modules/forms/services/FormFeedbackService.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      }
+
+      const csvGenerator = new FeedbackCsvGenerator(expectedNumResponses)
+
+      data.forEach((feedback) => {
+        csvGenerator.addLineFromFeedback(feedback)
+      })
+
+      csvGenerator.triggerFileDownload(`${formTitle}-${formId}-feedback.csv`)
+      return Promise.resolve()
```
Comment (mantariksh) [inline src/public/modules/forms/services/FormFeedbackService.ts:110]: you don't need this line, since the return type is `Promise<void>`

File: src/public/modules/forms/admin/directives/view-feedback.client.directive.js [hunk 12]
```diff
@@ -121,14 +114,13 @@ function viewFeedbackDirective(
...
-            formTitle: $scope.myform.title,
-          }
+          const formId = $scope.myform._id
+          const formTitle = $scope.myform.title
 
           $scope.csvDownloading = true
-          FormFeedback.downloadFeedback(params).finally(function () {
+          FormFeedback.downloadFeedback(formId, formTitle).finally(function () {
             $scope.csvDownloading = false
+            $scope.$apply()
```
Comment (mantariksh) [inline src/public/modules/forms/admin/directives/view-feedback.client.directive.js:123]: is the canonical approach here to use `$q` or `$scope`? I'm not sure

File: src/public/modules/forms/admin/directives/view-feedback.client.directive.js [hunk 13]
```diff
@@ -79,9 +74,7 @@ function viewFeedbackDirective(
         })
 
         $scope.createFeedbackTable = function (submissionCount) {
-          FormFeedback.getFeedback({
-            formId: $scope.myform._id,
-          }).then(
+          FormFeedback.getFeedback($scope.myform._id).then(
```
Comment (mantariksh) [inline src/public/modules/forms/admin/directives/view-feedback.client.directive.js:77]: this needs a `$q.when` or `$scope.apply()`

**Review by mantariksh — CHANGES_REQUESTED** [review]

File: src/app/modules/form/public-form/__tests__/public-form.controller.spec.ts [hunk 14]
```diff
@@ -126,9 +126,9 @@ describe('public-form.controller', () => {
         comment: MOCK_REQ.body.comment,
       })
       expect(mockRes.status).toHaveBeenCalledWith(200)
-      expect(mockRes.json).toHaveBeenCalledWith({
-        message: 'Successfully submitted feedback',
-      })
+      expect(mockRes.json).toHaveBeenCalledWith(
+        expect.objectContaining(mockFormFeedback),
```
[the remark this answers is not in the document]
Comment (tshuli) [reply src/app/modules/form/public-form/__tests__/public-form.controller.spec.ts]: thanks; no strong views although I thought it should be returned by default, so that we can display it as an acknowledgement if necessary. removed

File: src/types/form_feedback.ts [hunk 15]
```diff
@@ -2,16 +2,43 @@ import { Document, Model, QueryCursor } from 'mongoose'
...
   formId: IFormSchema['_id']
   rating: number
   comment?: string
 }
 
+export type FormFeedbackPostDto = Omit<IFormFeedback, 'formId'> & {
+  isPreview?: boolean
+}
+
+export type FormFeedbackResponseDto = IFormFeedback & {
```
[the remark this answers is not in the document]
Comment (tshuli) [reply src/types/form_feedback.ts]: thanks, shifted

File: src/public/modules/forms/admin/directives/view-feedback.client.directive.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (tshuli) [reply src/public/modules/forms/admin/directives/view-feedback.client.directive.js]: edited to use $q.when

File: src/public/modules/forms/admin/directives/view-feedback.client.directive.js [hunk 16]
```diff
@@ -121,14 +114,13 @@ function viewFeedbackDirective(
...
-          let params = {
-            formId: $scope.myform._id,
-            formTitle: $scope.myform.title,
-          }
+          const formId = $scope.myform._id
+          const formTitle = $scope.myform.title
 
           $scope.csvDownloading = true
-          FormFeedback.downloadFeedback(params).finally(function () {
+          FormFeedback.downloadFeedback(formId, formTitle).finally(function () {
```
[the remark this answers is not in the document]
Comment (tshuli) [reply src/public/modules/forms/admin/directives/view-feedback.client.directive.js]: edited

File: src/public/modules/forms/base/components/feedback-form.client.component.js [hunk 17]
```diff
@@ -31,19 +33,14 @@ function feedbackController(FormFeedback, Toastr) {
...
         },
-        function (error) {
-          console.error(error)
+        function (_error) {
```
[the remark this answers is not in the document]
Comment (tshuli) [reply src/public/modules/forms/base/components/feedback-form.client.component.js]: thanks, removed

File: src/public/modules/forms/helpers/CsvGenerator.ts [hunk 18]
```diff
@@ -29,7 +38,7 @@ module.exports = class CsvGenerator {
    * Insert raw data for a given row into the CSV file
    * @param {Array} rowData array of data to be inserted
```
[the remark this answers is not in the document]
Comment (tshuli) [reply src/public/modules/forms/helpers/CsvGenerator.ts]: edited

File: src/public/modules/forms/services/FormFeedbackService.ts (hunk 3 above)
Comment (tshuli) [reply src/public/modules/forms/services/FormFeedbackService.ts]: extracted and added tests

File: src/public/modules/forms/services/__tests__/FormFeedbackService.spec.ts (hunk 6 above)
Comment (tshuli) [reply src/public/modules/forms/services/__tests__/FormFeedbackService.spec.ts]: added tests

File: src/types/form_feedback.ts (hunk 1 above)
Comment (tshuli) [reply src/types/form_feedback.ts]: thanks, edited

File: src/public/modules/forms/base/components/feedback-form.client.component.js (hunk 2 above)
Comment (tshuli) [reply src/public/modules/forms/base/components/feedback-form.client.component.js]: thanks, changed to $q.when

File: src/public/modules/forms/helpers/CsvGenerator.ts (hunk 5 above)
Comment (tshuli) [reply src/public/modules/forms/helpers/CsvGenerator.ts]: thanks, edited

File: src/public/modules/forms/helpers/CsvGenerator.ts (hunk 7 above)
Comment (tshuli) [reply src/public/modules/forms/helpers/CsvGenerator.ts]: edited

File: src/public/modules/forms/services/FormFeedbackService.ts (hunk 9 above)
Comment (tshuli) [reply src/public/modules/forms/services/FormFeedbackService.ts]: shifted

File: src/public/modules/forms/services/FormFeedbackService.ts (hunk 3 above)
Comment (tshuli) [reply src/public/modules/forms/services/FormFeedbackService.ts]: shifted

File: src/public/modules/forms/services/FormFeedbackService.ts (hunk 11 above)
Comment (tshuli) [reply src/public/modules/forms/services/FormFeedbackService.ts]: removed

File: src/public/modules/forms/admin/directives/view-feedback.client.directive.js (hunk 12 above)
Comment (tshuli) [reply src/public/modules/forms/admin/directives/view-feedback.client.directive.js]: used $q.when throughout

File: src/app/modules/form/public-form/public-form.controller.ts (hunk 8 above)
Comment (tshuli) [reply src/app/modules/form/public-form/public-form.controller.ts]: removed as the response is no longer returned

File: src/public/modules/forms/services/FormFeedbackService.ts (hunk 10 above)
Comment (tshuli) [reply src/public/modules/forms/services/FormFeedbackService.ts]: I actually thought it looks ok now, but welcome suggestions!

File: src/app/modules/form/public-form/public-form.controller.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (tshuli) [reply src/app/modules/form/public-form/public-form.controller.ts]: thanks, changed back to previous implementation to return success message

File: src/public/modules/forms/helpers/__tests__/CsvGenerator.spec.ts [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+
+describe('CsvGenerator', () => {
+  afterAll(() => {
+    jest.clearAllMocks()
+  })
+
+  describe('Constructor', () => {
+    it('should initialise the instance properties correctly with defaults if params are not passed to constructor', () => {
+      // Arrange
+
```
Comment (karrui) [inline src/public/modules/forms/helpers/__tests__/CsvGenerator.spec.ts:20]: nit: remove empty blank line (also for the others)

File: src/public/modules/forms/helpers/__tests__/CsvGenerator.spec.ts [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+      expect(csv.startIdx).toEqual(2)
+      expect(csv.idx).toEqual(2)
+      expect(csv.records.length).toEqual(2)
+      expect(csv.records[0]).toEqual(UTF8_BYTE_ORDER_MARK)
+    })
+
+    it('should initialise the instance properties correctly if params are passed to constructor', () => {
+      // Arrange
+
+      const csv = new CsvGenerator(1, 2)
```
Comment (karrui) [inline src/public/modules/forms/helpers/__tests__/CsvGenerator.spec.ts:38]: should we assign the numbers in the constructor (`1`, `2`) to a constant so it is more obvious what we are looking at?

e.g.

```ts
const expectedNumberOfRecords = 1, numOfMetaDataRows = 2

...

expect(csv.expectedNumberOfRecords).toEqual(expectedNumberOfRecords)
expect(csv.numOfMetaDataRows).toEqual(numOfMetaDataRows)

// More obvious calculations for the other props too
expect(csv.startIdx).toEqual(expectedNumberOfRecords + numOfMetaDataRows + csv.numOfHeaderRows)
...

```

File: src/public/services/FormFeedbackService.ts [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+): Promise<void> => {
+  const expectedNumResponses = await countFeedback(formId)
+
+  return axios
+    .get<FormFeedbackResponseDto[]>(
+      `${ADMIN_FORM_ENDPOINT}/${formId}/feedback/download`,
+    )
+    .then(({ data }) => {
+      if (!data) {
+        return Promise.reject(new Error('Feedback size too large to download'))
```
Comment (karrui) [inline src/public/services/FormFeedbackService.ts:73]: is this the only reason? maybe add docs

File: src/types/api/form-feedback.ts [hunk 22]
```diff
@@ -0,0 +1,21 @@
+import { ProcessedFeedback } from '../form_feedback'
```
Comment (karrui) [inline src/types/api/form-feedback.ts:1]: do we want to standardize our file names? `src/types/form_feedback` is underscored but this file is hyphenated

**Review by karrui — APPROVED** [review]
lgtm with nits

File: src/app/modules/form/public-form/public-form.controller.ts [hunk 23]
```diff
@@ -66,7 +67,7 @@ const validateSubmitFormFeedbackParams = celebrate({
  */
 export const submitFormFeedback: RequestHandler<
   { formId: string },
-  ErrorDto | PrivateFormErrorDto,
+  FormFeedbackResponseDto | ErrorDto | PrivateFormErrorDto,
```
Comment (mantariksh) [inline src/app/modules/form/public-form/public-form.controller.ts:70]: this is no longer correct, right? the success case seems to now be `{ message: string }`

File: src/public/modules/forms/admin/directives/view-feedback.client.directive.js [hunk 24]
```diff
@@ -1,10 +1,12 @@
 'use strict'
 
+const FormFeedback = require('../../../../services/FormFeedbackService')
```
Comment (mantariksh) [inline src/public/modules/forms/admin/directives/view-feedback.client.directive.js:3]: tiny nit but can rename this `FormFeedbackService` for consistency with other new services

**Review by mantariksh — COMMENTED** [review]
looks good other than fixing the type on the controller. thanks for writing such detailed tests!

File: src/public/services/FormFeedbackService.ts (hunk 21 above)
Comment (tshuli) [reply src/public/services/FormFeedbackService.ts:73]: It was in the previous implementation, but I couldn't find any docs which suggest that `feedback size too large -> return empty object`. Replaced it with a more generic error message

File: src/types/api/form-feedback.ts (hunk 22 above)
Comment (tshuli) [reply src/types/api/form-feedback.ts:1]: hmm but within this folder `src/types/api`, `encrypt-submission.ts` is hyphenated too. ok to change it but should we have some standards documented to be consistent across?

File: src/app/modules/form/public-form/public-form.controller.ts (hunk 23 above)
Comment (tshuli) [reply src/app/modules/form/public-form/public-form.controller.ts:70]: thanks! edited

File: src/public/modules/forms/admin/directives/view-feedback.client.directive.js (hunk 24 above)
Comment (tshuli) [reply src/public/modules/forms/admin/directives/view-feedback.client.directive.js:3]: thanks! edited, and also for `feedback-form.client.component.js`

File: src/public/modules/forms/helpers/__tests__/CsvGenerator.spec.ts (hunk 20 above)
Comment (tshuli) [reply src/public/modules/forms/helpers/__tests__/CsvGenerator.spec.ts:38]: thanks, edited

**Comment by tshuli** [comment]
thanks @karrui and @mantariksh for reviewing

@mantariksh need an approval from you!

**Review by mantariksh — APPROVED** [review]

**Comment by mantariksh** [comment]
Closes #1837 also
